### PR TITLE
Deriving/manually implementing  `PartialOrd`/`Ord`/`Hash` instances wherever makes sense

### PR DESCRIPTION
**Pull request by chfanghr**

### CODE REVIEW

**Comment by chfanghr** [comment]
One thing I'm not sure is the default Ord instance for `BTreMap<T>`. Are they identical to the one found in haskell/plutarch?

**Review by bladyjoker — APPROVED** [review]
Awesome thanks! Let's just wait for @szg251 approval please

**Review by szg251 — APPROVED** [review]

File: src/v1/datum.rs [hunk 1]
```diff
@@ -24,7 +30,7 @@ impl IsPlutusData for DatumHash {
 }
 
 /// Piece of information associated with a UTxO encoded into a PlutusData type.
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
[the remark this answers is not in the document]
Comment (bladyjoker) [reply src/v1/datum.rs:33]: I don't think that's the point of Rust Hash, it's not for crypto purposes, it's for storing them in maps and sets and whatnot. Right?

File: src/v1/datum.rs [hunk 2]
```diff
@@ -8,11 +8,17 @@ use lbr_prelude::json::Json;
...
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
 #[cfg_attr(feature = "lbf", derive(Json))]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct DatumHash(pub LedgerBytes);
 
+impl std::hash::Hash for DatumHash {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(&self.0 .0)
+    }
+}
```
Comment (bladyjoker) [inline src/v1/datum.rs:20]: @szg251 did raise a good point, I don't think we should just pass the underlying Blake hash and just use `derive` on all of these as well. These serve a very different purpose imo.

Wdyt @chfanghr

**Review by bladyjoker — CHANGES_REQUESTED** [review]

File: src/v1/datum.rs (hunk 2 above)
Comment (chfanghr) [reply src/v1/datum.rs:20]: Well from the point of efficiency, it just feels weird to hash a hash.

Comment (bladyjoker) [reply src/v1/datum.rs:20]: From the point of semantics imo it's weird to not hash a hash. I don't really know what 'Hash' assumes, so using deriving throughout makes sense. There's several questions that pop out to me. Is there some length that's expected? Maybe `Hash` is more lenient and works internally with an collision expectation?

All these questions I don't really want to care about, and rather just use their internal generic derivation, which completely abolishes asking these questions.

On the point of efficiency; I generally deprioritize efficiency when it comes to uniformity. Anything requiring ad-hoc consideration is causing maintenance friction, which we can't really afford and must priority minimizing such concerns.

I hope that makes sense

Comment (chfanghr) [reply src/v1/datum.rs]: Understood, and it's done.

**Review by bladyjoker — APPROVED** [review]
Woot woot!! Thanks @chfanghr

**Pull request merged**